library/system.lisp

```diff
@@ -74,57 +66,110 @@ While the result will always contain microseconds, some implementations may retu
 
 (coalton-toplevel
 
-  (declare get-run-time (Unit -> UFix))
+  (declare get-run-time (Unit -> Integer))
   (define (get-run-time)
-    "Gets the run-time."
-    (lisp UFix ()
+    "Gets the run-time in internal time units. This is implementation specific: it may measure real time, run time, CPU cycles, or some other quantity.
+
+The difference between two successive calls to this function represents quantity accumulated during that period of time."
+    (lisp Integer ()
       (cl:get-internal-run-time)))
 
-  (declare get-real-time (Unit -> UFix))
+  (declare get-real-time (Unit -> Integer))
   (define (get-real-time)
-    "Gets the real-time."
-    (lisp UFix ()
+    "Gets the real-time in internal time units. The difference between two successive calls to this function represents the time that has elapsed."
+    (lisp Integer ()
       (cl:get-internal-real-time)))
 
-  #+sbcl
-  (declare get-bytes-consed (Unit -> UFix))
-  #+sbcl
-  (define (get-bytes-consed)
-    "Gets the number of bytes consed (only implemented for SBCL"
-    (lisp UFix ()
-      (sb-ext:get-bytes-consed)))
+  ;; (declare *internal-time-units-per-second* (Unit -> Integer))
+  (define  *internal-time-units-per-second*
+    "The number of internal time units per second. This is implementation specific."
+    (lisp Integer ()
+      cl:internal-time-units-per-second))
+
+  (declare time-units->seconds (Integer -> Double-Float))
+  (define (time-units->seconds t)
+    "Converts internal time units into seconds."
+    (real:inexact/ t *internal-time-units-per-second*))
+
+  (declare time-units->rounded-microseconds (Integer -> Integer))
+  (define (time-units->rounded-microseconds t)
+    "Converts internal time units into microseconds."
+    (real:round/ (* 1000000 t)
+                 *internal-time-units-per-second*))
+
+  (declare monotonic-bytes-consed (Unit -> (Optional Integer)))
+  (define (monotonic-bytes-consed)
+    "Returns the number of bytes consed since some unspecified point in time.
+
+The difference between two successive calls to this function represents the number of bytes consed in that period of time."
+    #+sbcl
+    (Some (lisp Integer ()
+            (sb-ext:get-bytes-consed)))
+    #-sbcl
+    None)
+
+  ;;
+  ;;
+  ;;
+
+  (declare time ((Unit -> :a) -> (Tuple :a Integer)))
+  (define (time f)
+    "Run the thunk `f` and return a tuple containing its value along with the run time in microseconds.
+
+While the result will always contain microseconds, some implementations may return a value rounded to less precision (e.g., rounded to the nearest second or millisecond)."
+    (let start = (get-run-time))
+    (let value = (f))
+    (let end   = (get-run-time))
+    (Tuple value (time-units->rounded-microseconds (- end start))))
+
+  (declare space ((Unit -> :a) -> (Tuple :a (Optional Integer))))
+  (define (space f)
+    "Run the thunk `f` and return a tuple containing its value along with the number of bytes consed.
```

Comment:
"...approximate number of bytes consed during the course of executing `f`. The amount of space used may be peculiar to the implementation, such as rounding to certain page boundaries."

Reply:
We might even just want to return a fraction. Why approximate it for them?